applications/doodle/src/doodle_listener.erl

```diff
@@ -286,66 +289,55 @@ account_from_number(#{phone_number := KNumber, request := JObj} = Map) ->
         {'ok', Id, OwnerId} ->
-            Props = props:filter_undefined([{?CCV(<<"Authorizing-Type">>), <<"device">>}
-                                           ,{?CCV(<<"Authorizing-ID">>), Id}
-                                           ,{?CCV(<<"Owner-ID">>), OwnerId}
+            Props = props:filter_undefined([{?CV(<<"Authorizing-Type">>), <<"device">>}
+                                           ,{?CV(<<"Authorizing-ID">>), Id}
+                                           ,{?CV(<<"Owner-ID">>), OwnerId}
                                            ]),
-            JObj1 = kz_json:delete_keys([?CCV(<<"Authorizing-Type">>)
-                                        ,?CCV(<<"Authorizing-ID">>)
+            JObj1 = kz_json:delete_keys([?CV(<<"Authorizing-Type">>)
+                                        ,?CV(<<"Authorizing-ID">>)
                                         ], JObj),
             Map#{request => kz_json:set_values(Props, JObj1)};
         {'error', _} -> Map
     end;
 lookup_mdn(Map) -> Map.
 
--spec set_default_headers(map()) -> map().
-set_default_headers(#{request:= SmsReq, account_id := _AccountId} = Map) ->
-    Map#{request => kz_json:set_values(kz_api:default_headers(?APP_NAME, ?APP_VERSION), SmsReq)};
-set_default_headers(Map) -> Map.
-
--spec set_account_id(map()) -> map().
-set_account_id(#{request:= SmsReq, account_id := AccountId} = Map) ->
-    Map#{request => kz_api_sms:set_account_id(SmsReq, AccountId)};
-set_account_id(Map) -> Map.
-
--spec set_caller_id(map()) -> map().
-set_caller_id(#{request:= SmsReq
-               ,account_id := AccountId
-               ,authorizing_id := AuthorizingId
-               } = Map) ->
-    case kz_endpoint:get(AuthorizingId, AccountId) of
-        {'error', Error} ->
-            lager:warning("unable to get endpoint ~s/~s : ~p", [AuthorizingId, AccountId, Error]),
-            Map;
-        {'ok', EP} ->
-            CID = kzd_devices:caller_id(EP, kz_json:new()),
-            From = kzd_caller_id:internal_number(CID, kz_api_sms:from(SmsReq)),
-            Map#{request => kz_api_sms:set_from(SmsReq, From)}
-    end;
-set_caller_id(Map) -> Map.
-
--spec create_im(map()) -> map().
-create_im(#{request:= SmsReq, account_id := _AccountId} = Map) ->
-    IM = kapps_im:from_payload(SmsReq),
+-spec create_im(context()) -> context().
+create_im(#{request:= SmsReq, account_id := AccountId} = Map) ->
+    Funs = [{fun kapps_im:set_application_name/2, ?APP_NAME}
+           ,{fun kapps_im:set_application_version/2, ?APP_VERSION}
+           ,{fun kapps_im:set_account_id/2, AccountId}
+           ,fun maybe_update_sender/1
+           ],
+    IM = kapps_im:exec(Funs, kapps_im:from_payload(SmsReq)),
     Map#{fetch_id => kapps_im:message_id(IM)
         ,message_id => kapps_im:message_id(IM)
         ,im => IM
         };
 create_im(Map) -> Map.
 
+-spec maybe_update_sender(kapps_im:im()) -> kapps_im:im().
+maybe_update_sender(Im) ->
+    case kapps_im:endpoint(Im) of
```

Comment:
Dialyzer doesn't think `undefined` will be reached, since the Im record defaults to empty JSON object. Would you prefer to test for empty JObj and return Im unmodified or just bind EP and update the From?